### PR TITLE
[WIP] Widgetastic conversion of cfme.cloud.instance

### PR DESCRIPTION
**Pull request by mshriver**
Conversion of cfme.cloud.instance class to widgetastic views/forms.

Nested the forms in the views for each navigation destination. Not all of the navigation destinations are used by tests.

I do not plan to fix all failing tests within this PR, only to convert the navigation/form/widget use and to fix any failures that are a result of that conversion.

{{ pytest: cfme/tests/cloud/test_instance_power_control.py cfme/tests/cloud/test_provisioning.py --long-running --use-provider rhos7-ga --use-provider ec2west --use-provider azure }}

## PRT

Fixing deferred_verpick errors in tests with PR #4544 
Other failures/skips that I'm seeing aren't related to this migration but are in effected tests so they show up here.

### CODE REVIEW

**Comment by dajoRH** [comment]
Lint report for commit 55da7f8d885f150c6d00adfe0ca9e37c0d454aeb:

`cfme/provisioning.py`:
- :red_circle: Line 45:11: **N801** *class names should use CapWords convention*
- :red_circle: Line 54:11: **N801** *class names should use CapWords convention*
- :red_circle: Line 61:11: **N801** *class names should use CapWords convention*
- :red_circle: Line 78:11: **N801** *class names should use CapWords convention*
- :red_circle: Line 103:11: **N801** *class names should use CapWords convention*
- :red_circle: Line 116:11: **N801** *class names should use CapWords convention*
- :red_circle: Line 121:11: **N801** *class names should use CapWords convention*
- :red_circle: Line 134:11: **N801** *class names should use CapWords convention*
- :red_circle: Line 160:11: **N801** *class names should use CapWords convention*

Hmmm
:frowning: You might like to check the lint of your code :point_up: 


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit a99ebd1a96191fe31be85a086594bb43d50ace7d:

`cfme/provisioning.py`:
- :red_circle: Line 45:11: **N801** *class names should use CapWords convention*
- :red_circle: Line 54:11: **N801** *class names should use CapWords convention*
- :red_circle: Line 61:11: **N801** *class names should use CapWords convention*
- :red_circle: Line 78:11: **N801** *class names should use CapWords convention*
- :red_circle: Line 103:11: **N801** *class names should use CapWords convention*
- :red_circle: Line 116:11: **N801** *class names should use CapWords convention*
- :red_circle: Line 121:11: **N801** *class names should use CapWords convention*
- :red_circle: Line 134:11: **N801** *class names should use CapWords convention*
- :red_circle: Line 160:11: **N801** *class names should use CapWords convention*

Hmmm
:frowning: You might like to check the lint of your code :point_up: 


**No commit flaws detected**.

*CFME QE Bot*

File: cfme/cloud/instance/__init__.py [hunk 1]
```diff
@@ -38,33 +35,116 @@
...
+        tree = ManageIQTree()
+
+    @View.nested
+    class instances(Accordion): # noqa
+        ACCORDION_NAME = 'Instances'
+        tree = ManageIQTree()
+
+    @View.nested
+    class images(Accordion): # noqa
+        ACCORDION_NAME = 'Images'
```
Comment (mfalesni) [inline cfme/cloud/instance/__init__.py]: If the accordion text is corresponding to the titleized class name, then you don't need to set the ACCORDION_NAME. Not that I would dislike this usage, but the possibility is there :)

File: cfme/cloud/instance/__init__.py [hunk 2]
```diff
@@ -38,33 +35,116 @@
...
+    def is_displayed(self):
+        # TODO: Match against image selected in Catalog tab
+        return False
+
+    def before_fill(self, values):
+        # Fill first part of form 'manually' here
+        # fingers crossed this works like I'm planning
+        try:
+            row = self.image_table.row(Name=self.context['object'].template_name,
+                                       Provider=self.context['object'].provider.name)
```
Comment (mfalesni) [inline cfme/cloud/instance/__init__.py]: When we're using the keywords, I would stick to the `snake_cased` with all-lowercase to keep in line with python style.

File: cfme/cloud/instance/__init__.py [hunk 3]
```diff
@@ -382,13 +435,15 @@ def step(self):
 
 @navigator.register(Instance, 'Provision')
 class InstanceProvision(CFMENavigateStep):
-    prerequisite = NavigateToSibling('All')
+    VIEW = InstanceProvisionView
 
-    def am_i_here(self):
-        return match_page(summary='Provision Instances - Select an Image')
+    def prerequisite(self, *args, **kwargs):
+        self.prereq = navigate_to(Instance, 'All')
```
Comment (mfalesni) [inline cfme/cloud/instance/__init__.py]:
```
utils/appliance/implementations/ui.py
353:            self.parent_view = self.prerequisite()
```

File: cfme/cloud/instance/__init__.py (hunk 1 above)
Comment (mshriver) [inline cfme/cloud/instance/__init__.py]: I see - wasn't sure about the more_complex_names, and how they would be parsed/matched. Personally I like setting it explicitly, its more clear to me (at my current level of proficiency) what the nested view class is for.  Not against removing if we want to keep a consistent style.

File: cfme/cloud/instance/__init__.py (hunk 2 above)
Comment (mshriver) [inline cfme/cloud/instance/__init__.py]: Understood - didn't dive too deep into widgetastic Table/row classes, just saw that row is `attributized`, thought I had to match the column name capitalization.

File: cfme/cloud/instance/__init__.py (hunk 3 above)
Comment (mshriver) [inline cfme/cloud/instance/__init__.py]: :man_facepalming: Then I can call `self.parent_view.lifecycle` perfect.

File: cfme/provisioning.py [hunk 4]
```diff
@@ -35,6 +40,137 @@ def select_security_group(sg):
     sel.sleep(1)
 
 
+class ProvisioningForm(BaseLoggedInPage):
+    @View.nested
+    class request(Tab):  # noqa
+        TAB_NAME = 'Request'
+        email = Input('requester__owner_email')
```
Comment (mshriver) [inline cfme/provisioning.py]: @mfalesni @izapolsk  When I try to fill the form with a nested dictionary of values, this field's fill raises an exception at widget.TextInput.fill(). self.value triggers the exception for NoSuchElement.

While debugging, the form was loaded, and manually searching for the locator stored in the TextInput object showed that the xpath was correct.

Comment (mfalesni) [inline cfme/provisioning.py]: Look for the definition of Tab class and add `INDIRECT = True`.

@mshriver  I have it in my AE PR but if anyone beats me to it, I am okay with that :)

INDIRECT causes to ignore the root locator for given view class, which is required for Tab as the ROOT contains only the tab on top but the actual content is located somewhere else :)

Comment (mfalesni) [inline cfme/provisioning.py]: Oh, and please, use the keyword args for Inputs and related classes (`Input(name='...')`)

Comment (mshriver) [inline cfme/provisioning.py]: Used it below, think I hadn't realized it needed to be specified yet here. thanks.

**Comment by dajoRH** [comment]
Lint report for commit 89185daa939c6b973b9bb4235a133352c408d4f2:

`cfme/cloud/instance/ec2.py`:
- :red_circle: Line 2:1: **F401** *'widgetastic_patternfly.FlashMessages' imported but unused*
- :red_circle: Line 4:1: **F401** *'cfme.BaseLoggedInPage' imported but unused*
- :red_circle: Line 7:1: **F401** *'cfme.services.requests.RequestView' imported but unused*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

File: cfme/provisioning.py [hunk 5]
```diff
@@ -35,6 +40,136 @@ def select_security_group(sg):
     sel.sleep(1)
 
 
+class ProvisioningForm(BaseLoggedInPage):
```
Comment (mshriver) [inline cfme/provisioning.py]: @izapolsk I'm thinking I should maybe pull this into a separate PR that we can review much faster that all the instance stuff. thoughts?

Comment (izapolsk) [inline cfme/provisioning.py]: @mshriver , good idea. I also think of splitting my PR to several ones.

Comment (mshriver) [inline cfme/provisioning.py]: @izapolsk @psav done, PR #4031

**Comment by dajoRH** [comment]
Lint report for commit 7505ce3da042522533f7c3ce53ffe812085e1921:

`cfme/cloud/instance/azure.py`:
- :bangbang: Line 33:21: **E241** *multiple spaces after ','*
- :bangbang: Line 33:68: **E241** *multiple spaces after ','*
- :red_circle: Line 55:26: **F821** *undefined name 'email'*
- :red_circle: Line 56:31: **F821** *undefined name 'first_name'*
- :red_circle: Line 57:30: **F821** *undefined name 'last_name'*
- :red_circle: Line 68:34: **F821** *undefined name 'instance_type'*

`cfme/cloud/instance/ec2.py`:
- :bangbang: Line 28:21: **E241** *multiple spaces after ','*
- :red_circle: Line 50:26: **F821** *undefined name 'email'*
- :red_circle: Line 51:31: **F821** *undefined name 'first_name'*
- :red_circle: Line 52:30: **F821** *undefined name 'last_name'*
- :red_circle: Line 63:34: **F821** *undefined name 'instance_type'*

`cfme/cloud/instance/gce.py`:
- :bangbang: Line 32:21: **E241** *multiple spaces after ','*
- :red_circle: Line 53:26: **F821** *undefined name 'email'*
- :red_circle: Line 54:31: **F821** *undefined name 'first_name'*
- :red_circle: Line 55:30: **F821** *undefined name 'last_name'*

`cfme/cloud/instance/openstack.py`:
- :red_circle: Line 65:26: **F821** *undefined name 'email'*
- :red_circle: Line 66:31: **F821** *undefined name 'first_name'*
- :red_circle: Line 67:30: **F821** *undefined name 'last_name'*
- :red_circle: Line 73:34: **F821** *undefined name 'cloud_network'*
- :red_circle: Line 76:34: **F821** *undefined name 'instance_type'*
- :bangbang: Line 82:9: **E303** *too many blank lines (3)*

Hmmm
:frowning: You might like to check the lint of your code :point_up: 


**No commit flaws detected**.

*CFME QE Bot*

**Comment by mshriver** [comment]
Opening back up to bring up-to-date with some of the newest widgets.  Definitely needs some cleanup.

**Comment by dajoRH** [comment]
Lint report for commit 6e03bdaefa6f7c308418feb9f81c93ee5567f9a9:

`cfme/cloud/instance/__init__.py`:
- :bangbang: Line 430:5: **E303** *too many blank lines (2)*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit ef7f21ee61e851e4ad11408f8db982af3b7af7e5:

`cfme/cloud/instance/__init__.py`:
- :red_circle: Line 7:1: **F401** *'widgetastic_manageiq.SummaryFormItem' imported but unused*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by mshriver** [comment]
Blocked by #4383

File: cfme/cloud/instance/__init__.py [hunk 6]
```diff
@@ -192,318 +224,558 @@ def find_quadicon(self, *args, **kwargs):
...
         # If a filter was applied, it will persist through navigation and needs to be cleared
         if sel.is_displayed(search_box.clear_advanced_search):
             logger.debug('Clearing advanced search filter')
             sel.click(search_box.clear_advanced_search)
-        accordion.tree('Instances by Provider', 'Instances by Provider', self.obj.provider.name)
+
+
+class AddFloatIPView(CloudInstanceView):
+    @View.nested
+    class form(ManageIQForm):
```
Comment (mshriver) [inline cfme/cloud/instance/__init__.py]: @izapolsk I had separate views and forms before, but it was much too cluttered with far too many module level classes.

I've collapsed the forms into nested views, what do you think of this structure?  Note `ManageIQForm` is new, and if the team likes it I can split into a separate PR.

**Comment by dajoRH** [comment]
Lint report for commit b52bbc63745cf14d3b4864ba14aac4d6e2900729:

`cfme/cloud/instance/__init__.py`:
- :red_circle: Line 306:11: **N801** *class names should use CapWords convention*
- :red_circle: Line 333:11: **N801** *class names should use CapWords convention*
- :red_circle: Line 427:11: **N801** *class names should use CapWords convention*
- :red_circle: Line 453:11: **N801** *class names should use CapWords convention*
- :red_circle: Line 482:11: **N801** *class names should use CapWords convention*
- :red_circle: Line 503:11: **N801** *class names should use CapWords convention*
- :red_circle: Line 525:11: **N801** *class names should use CapWords convention*
- :red_circle: Line 549:11: **N801** *class names should use CapWords convention*
- :red_circle: Line 571:11: **N801** *class names should use CapWords convention*
- :red_circle: Line 609:11: **N801** *class names should use CapWords convention*
- :red_circle: Line 630:11: **N801** *class names should use CapWords convention*
- :red_circle: Line 676:11: **N801** *class names should use CapWords convention*
- :red_circle: Line 702:11: **N801** *class names should use CapWords convention*
- :red_circle: Line 724:11: **N801** *class names should use CapWords convention*
- :red_circle: Line 761:11: **N801** *class names should use CapWords convention*

Hmmm
:frowning: You might like to check the lint of your code :point_up: 


**No commit flaws detected**.

*CFME QE Bot*

File: widgetastic_manageiq.py [hunk 7]
```diff
@@ -1589,3 +1589,19 @@ class chart(TimelinesChart):  # NOQA
     @property
     def is_displayed(self):
         return self.title.text == 'Timelines'
+
```
Comment (mshriver) [inline widgetastic_manageiq.py]: @mfalesni @izapolsk @otsuman @sshveta You have all done some widgets now, I just prototyped this generic form view to eliminate duplicate button definitions in the various forms in cloud.instances and that will be in almost all forms for the application.

You can see the use above in cloud.instance, I have nested a view that inherits this form, and then defined the unique form widgets.

Thoughts? Any other prototypes/solutions for this at the moment that I'm missing?

Comment (mshriver) [inline widgetastic_manageiq.py]: Obviously this isn't all the generic buttons, there's 'Save' and probably some more that we'll want, maybe for the search forms (load, delete)

Comment (sshveta) [inline widgetastic_manageiq.py]: My :+1:  I like the nested view , easy to understand which form belongs to which view and ofcourse no need to define buttons

**Comment by dajoRH** [comment]
Lint report for commit c9ae00a218697e694c7d0f93b1f071a16efcf8c6:

`cfme/cloud/instance/__init__.py`:
- :red_circle: Line 571:11: **N801** *class names should use CapWords convention*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit a60a53d232cad32d82ff4ff1d618c42811d3ed64:

`cfme/cloud/instance/__init__.py`:
- :red_circle: Line 591:11: **N801** *class names should use CapWords convention*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

File: cfme/cloud/instance/__init__.py [hunk 8]
```diff
@@ -192,318 +224,594 @@ def find_quadicon(self, *args, **kwargs):
...
+
+@navigator.register(Instance, 'AddFloatingIP')
+class AddFloatingIP(CFMENavigateStep):
+    VIEW = AddFloatIPView
+    prerequisite = NavigateToSibling('Details')
+
+    def step(self):
+        if version.current_version() >= '5.7':
+            try:
+                self.prerequisite_view.configuration.item_select(
```
Comment (mshriver) [inline cfme/cloud/instance/__init__.py]: I believe its only for rhos instances

File: cfme/cloud/instance/__init__.py [hunk 9]
```diff
@@ -192,318 +224,594 @@ def find_quadicon(self, *args, **kwargs):
...
+                    'Associate a Floating IP with this Instance')
+            except NoSuchElementException:
+                raise DestinationNotFound('Add Floating IP option not available for instance')
+        else:
+            raise DestinationNotFound('Floating IP assignment not available for appliance version')
+
+
+class AttachVolumeView(CloudInstanceView):
+    @View.nested
```
Comment (mshriver) [inline cfme/cloud/instance/__init__.py]: This made more sense to me, because not all widgets in the view are part of the form, and some views have a form and a display (like timelines). Checking out toolbar view structure from your other comment.

File: cfme/cloud/instance/__init__.py [hunk 10]
```diff
@@ -192,318 +224,594 @@ def find_quadicon(self, *args, **kwargs):
...
+                                                             'Instance')
+        except NoSuchElementException:
+            raise DestinationNotFound('Detach Cloud Volume option not available for instance')
+
+
+class EditView(CloudInstanceView):
+    @View.nested
+    class form(View):  # noqa
+        custom_identifier = Input(name='custom_1')
+        description = Input(name='')
```
Comment (mshriver) [inline cfme/cloud/instance/__init__.py]: fixing, thx

File: cfme/cloud/instance/__init__.py [hunk 11]
```diff
@@ -101,25 +96,60 @@ def __init__(self, name, provider, template_name=None, appliance=None):
...
+        if reset:
+            view.form.reset_button.click()
+            return
         else:
-            return InfoBlock.text(*kwargs.get('properties'))
+            button = view.form.cancel_button if cancel else view.form.submit_button
+            button.click()
 
     def on_details(self, force=False):
         """A function to determine if the browser is already on the proper instance details page.
```
Comment (mshriver) [inline cfme/cloud/instance/__init__.py]: its still called dynamically through common as far as I can tell, so until those dynamic uses are removed it has to stay. Just returns DetailsView.is_displayed  for now?

File: cfme/cloud/instance/__init__.py [hunk 12]
```diff
@@ -101,25 +96,60 @@ def __init__(self, name, provider, template_name=None, appliance=None):
...
+            cancel: Boolean, cancel the form submission
+            reset Boolean, reset form after fill - returns immediately after reset
+        Note:
+            The edit form contains a 'Reset' button - if this is c
+        """
+        view = navigate_to(self, 'Edit')
+        # form is the view's parent
+        view.form.fill(values)
+        if reset:
+            view.form.reset_button.click()
```
Comment (mshriver) [inline cfme/cloud/instance/__init__.py]: I would think in the case where the caller wants the form reset after fill, they may be testing reset and this should return immediately?

I did add utils.Updateable to the Instance class

File: cfme/cloud/instance/__init__.py [hunk 13]
```diff
@@ -192,318 +224,594 @@ def find_quadicon(self, *args, **kwargs):
...
+        except NoSuchElementException:
+            raise DestinationNotFound('Evacuate option not available for instance')
 
 
-@navigator.register(Instance, 'AttachCloudVolume')
-class InstanceAttachVolume(CFMENavigateStep):
+class ManagePoliciesView(CloudInstanceView):
+    @View.nested
+    class form(View):  # noqa
+        policy_profiles = CheckboxSelect('protectbox')
```
Comment (izapolsk) [inline cfme/cloud/instance/__init__.py]: please test this because CheckboxSelect didn't work for me in provider's manage policies form.

Comment (mshriver) [inline cfme/cloud/instance/__init__.py]: Will do, thanks!

File: cfme/cloud/instance/__init__.py [hunk 14]
```diff
@@ -1,80 +1,72 @@
...
-        ('template_table', Table('//div[@id="pre_prov_div"]//table')),
-        ('cancel_button', form_buttons.cancel)
-    ]
-)
+from utils.wait import wait_for
 
-manage_policies_tree = CheckboxTree("//div[@id='protect_treebox']/ul")
 
-list_table = PagedTable(table_locator="//div[@id='list_grid']//table")
+class CloudInstanceView(BaseLoggedInPage):
```
Comment (mshriver) [inline cfme/cloud/instance/__init__.py]: @izapolsk I would like to make this refactor in conjunction with the common views for virtual machines and instances.

File: cfme/cloud/instance/__init__.py (hunk 8 above)
Comment (mshriver) [inline cfme/cloud/instance/__init__.py]: Tested with RHOS, the Floating IP field is just widget.Select, thought it was BootstrapSelect.

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit b3fd23ae7af5374d66f511a35e03b72157d9ed1a:

`cfme/cloud/instance/azure.py`:
- :red_circle: Line 3:1: **F401** *'cfme.common.vm.VM' imported but unused*

`cfme/cloud/instance/ec2.py`:
- :red_circle: Line 2:1: **F401** *'cfme.common.vm.VM' imported but unused*

`cfme/cloud/instance/gce.py`:
- :red_circle: Line 3:1: **F401** *'cfme.common.vm.VM' imported but unused*
- :bangbang: Line 41:5: **E303** *too many blank lines (2)*

`cfme/cloud/instance/openstack.py`:
- :red_circle: Line 2:1: **F401** *'cfme.common.vm.VM' imported but unused*

Hmmm
:frowning: You might like to check the lint of your code :point_up: 


**No commit flaws detected**.

*CFME QE Bot*

File: cfme/cloud/instance/__init__.py [hunk 15]
```diff
@@ -191,318 +223,604 @@ def find_quadicon(self, *args, **kwargs):
...
-            return sel.click(list_table.find_row_by_cell_on_all_pages(
-                {'Name': self.obj.name,
-                 'Provider': self.obj.provider.name}))
-        except (NameError, TypeError):
-            logger.warning('Exception caught, could not match instance with name and provider')
+            row = self.prerequisite_view.paginator.find_row_on_pages(
+                self.prerequisite_view.instance_table,
+                name=self.obj.name,
+                provider=self.obj.provider.name)
+            self.log_message('Matched table row on instance name and provider')
```
Comment (izapolsk) [inline cfme/cloud/instance/__init__.py]: :+1:

File: cfme/cloud/instance/__init__.py [hunk 16]
```diff
@@ -1,79 +1,72 @@
...
+        ACCORDION_NAME = 'Instances'
+        tree = ManageIQTree()
+
+    @View.nested
+    class images(Accordion):  # noqa
+        ACCORDION_NAME = 'Images'
+        tree = ManageIQTree()
+
+
+@VM.register_for_provider_type('cloud')
```
Comment (izapolsk) [inline cfme/cloud/instance/__init__.py]: REQUIRED: these things were removed lately by psav during transition to entrypoints. Please make sure you correctly merged with PR #4500.

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

File: cfme/cloud/instance/__init__.py [hunk 17]
```diff
@@ -127,12 +157,9 @@ def on_details(self, force=False):
...
             for an instance on the wrong provider.
         """
-
-        if details_page_check(name=self.name, provider=self.provider):
-            return True
-        elif not force:
-            return False
-        elif force:
+        if not force:
+            return self.browser.create_view(DetailsView).is_displayed()
```
Comment (mfalesni) [inline cfme/cloud/instance/__init__.py]: is_displayed is a property

File: cfme/cloud/instance/__init__.py [hunk 18]
```diff
@@ -100,25 +95,60 @@ def __init__(self, name, provider, template_name=None, appliance=None):
...
+            recursive_update(form_values, {'environment': {'security_groups': first_group}})
+
+        view.form.fill(form_values)
+
+        if cancel:
+            view.form.cancel_button.click()
+            # Redirects to Instance All, but flash is defined in BaseLoggedInPage so use directly
+            view = self.browser.create_view(BaseLoggedInPage)
+            wait_for(func=view.flash.assert_success_message, func_arg=self.PROVISION_CANCEL,
+                     fail_condition=[], timeout=10, delay=2, message='wait for flash success')
```
Comment (mfalesni) [inline cfme/cloud/instance/__init__.py]: There is no such arg as func_arg, you probably wanted to pass a list to func_args.

Flash messages widget does not return anything, always returns None then.

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by mshriver** [comment]
Closing in favor of smaller PR's with a few views at a time.
